### PR TITLE
Enhance dep status to support detailed output

### PR DESCRIPTION
**Pull request by zjs**
Enhance `dep status` to support detailed output for the tabular, JSON, and `text/template` output formats. For each of these formats, the detailed output includes `source` repository information and a list of packages (instead of simply the count). For the JSON and `text/template` output formats, this detailed view separates branch, version, and revision information (for both locked and latest) to simplify programmatic processing of the output. For the JSON format, solve metadata is included in the output.

Additionally, this sets the stage for an all-at-once `text/template` outputter (like the JSON outputter, instead of the current line-by-line approach preserved from the basic `text/template` outputter) to allow a template to additionally process metadata information.

### CODE REVIEW

File: cmd/dep/status.go [hunk 1]
```diff
@@ -152,6 +158,29 @@ func (out *tableOutput) BasicLine(bs *BasicStatus) error {
...
+	_, err := fmt.Fprintf(out.w, "PROJECT\tSOURCE\tCONSTRAINT\tVERSION\tREVISION\tLATEST\tPKGS USED\n")
+	return err
+}
+
+func (out *tableOutput) DetailFooter(metadata *dep.SolveMeta) error {
+	return out.BasicFooter()
+}
+
+func (out *tableOutput) DetailLine(ds *DetailStatus) error {
+	_, err := fmt.Fprintf(out.w,
```
Comment (akutz) [inline cmd/dep/status.go:174]: Hi @zjs,

Are you familiar with [`text/tabwriter`](https://golang.org/pkg/text/tabwriter/)?

File: cmd/dep/status.go [hunk 2]
```diff
@@ -865,23 +1024,31 @@ func runStatusAll(ctx *dep.Ctx, out outputter, p *dep.Project, sm gps.SourceMana
...
 		}
 
-		// A map of ProjectRoot and *BasicStatus. This is used in maintain the
-		// order of BasicStatus in output by collecting all the BasicStatus and
-		// then using them in order.
-		bsMap := make(map[string]*BasicStatus)
-		for bs := range bsCh {
-			bsMap[bs.ProjectRoot] = bs
-		}
+		if cmd.detail {
```
Comment (akutz) [inline cmd/dep/status.go:1044]: Hi @zjs,

I would include `return false, errCount, err` at the end of this `if cmd.detail {` block and then get rid of the `else` block below and return again at the end. This removes a level of indentation without making things any more difficult to follow.

File: cmd/dep/status.go [hunk 3]
```diff
@@ -652,13 +792,27 @@ func (bs *BasicStatus) marshalJSON() *rawStatus {
...
+}
+
 // MissingStatus contains information about all the missing packages in a project.
 type MissingStatus struct {
 	ProjectRoot     string
 	MissingPackages []string
 }
 
-func runStatusAll(ctx *dep.Ctx, out outputter, p *dep.Project, sm gps.SourceManager) (hasMissingPkgs bool, errCount int, err error) {
+func (cmd *statusCommand) runStatusAll(ctx *dep.Ctx, out outputter, p *dep.Project, sm gps.SourceManager) (hasMissingPkgs bool, errCount int, err error) {
```
Comment (akutz) [inline cmd/dep/status.go:832]: Hi @zjs,

I think we should try to avoid refactoring things to our preference if possible. There's no real reason that this function needs to be assigned to `*statusCmd`, is there? I'm not a fan of the changes to the channel work below either, because I wonder if it doesn't make sense to just include the detail status as a field of basic status and again, reduce changes to what is truly necessary. Then again, we can always change it later if there is pushback from the project maintainers.

File: cmd/dep/status.go [hunk 4]
```diff
@@ -956,6 +1123,48 @@ outer:
 	return hasMissingPkgs, 0, errInputDigestMismatch
 }
 
+// basicOutputAll takes an outputter, a sorted project list, and a map of ProjectRoot to *BasicStatus and
```
Comment (akutz) [inline cmd/dep/status.go]: Hi @zjs,

What do you mean a sorted project list? Is that comment copied from elsewhere? We should maintain the order the solver provides.

File: cmd/dep/status.go [hunk 5]
```diff
@@ -956,6 +1123,48 @@ outer:
...
+	}
+
+	if footerErr := out.BasicFooter(); footerErr != nil {
+		return footerErr
+	}
+
+	return nil
+}
+
+// detailOutputAll takes an outputter, a sorted project list, and a map of ProjectRoot to *DetailStatus and
```
Comment (akutz) [inline cmd/dep/status.go]: Hi @zjs,

Please see above regarding sort order.

File: cmd/dep/status.go [hunk 6]
```diff
@@ -973,6 +1182,40 @@ func formatVersion(v gps.Version) string {
...
+func formatDetailVersion(v gps.Version, r gps.Revision) rawDetailVersion {
+	if v == nil {
+		return rawDetailVersion{
+			Revision: r.String(),
+		}
+	}
+	switch v.Type() {
+	case gps.IsBranch:
+		return rawDetailVersion{
+			Branch:   v.String(),
```
Comment (akutz) [inline cmd/dep/status.go:1212]: Hi @zjs,

Does this include the `"branch BRANCHNAME"` format when executed like this? If so, please check for that and reduce to just `BRANCHNAME`.

**Review by akutz — CHANGES_REQUESTED** [review]
Hi @zjs,

This looks really, really good! There are some places I requested feedback, and I'm ready to respond as soon as you do :)

File: cmd/dep/status.go (hunk 6 above)
Comment (zjs) [reply cmd/dep/status.go:1212]: This won't include the "branch" string; that's added by the `formatVersion` function above, which this method is called instead of. I'd rename that something like `formatBasicVersion`, but it's also used for the `old` and `missing` codepaths.

Comment (akutz) [reply cmd/dep/status.go:1212]: Hi @zjs,

Gotcha. Just checking since we had that conversation earlier.

File: cmd/dep/status.go (hunk 2 above)
Comment (akutz) [reply cmd/dep/status.go:1044]: Not a huge deal though. Just my personal preference. Neither way is more right than the other.

File: cmd/dep/status.go (hunk 4 above)
Comment (zjs) [reply cmd/dep/status.go]: This is based on existing comments from where the argument is defined.

https://github.com/akutz/dep/blob/5f52a57c47b1a9e5714736c0d266bd130154590c/cmd/dep/status.go#L872-L878

Essentially, I'm just trying to express that this function assumes that the supplied input order is what you'd like the output order to be. I'll reword this to actually say that.

Comment (akutz) [reply cmd/dep/status.go]: Hi @zjs,

Understood. I just didn't see anywhere you added new sorting, but I wanted to make sure since the order is so specific with which to begin.

File: cmd/dep/status.go (hunk 3 above)
Comment (akutz) [reply cmd/dep/status.go:832]: Hi @zjs,

Regarding the changes to the channel workflow. I think I'm leaning towards what I said above re: pushback. Let's keep your design, and if the dep folks don't like it, we can revert it to the previous incarnation and change the composite object to a field in `BasicStatus`.

Comment (zjs) [reply cmd/dep/status.go:832]: > I think we should try to avoid refactoring things to our preference if possible. There's no real reason that this function needs to be assigned to *statusCmd, is there?

`runStatusAll` needs to be able to check `cmd.detail` to know whether to output "basic" or "detailed" information.

That's checked here:

https://github.com/akutz/dep/blob/5f52a57c47b1a9e5714736c0d266bd130154590c/cmd/dep/status.go#L1044

Is there a different way you'd recommend conveying that?

> Regarding the changes to the channel workflow. I think I'm leaning towards what I said above re: pushback. Let's keep your design, and if the dep folks don't like it, we can revert it to the previous incarnation and change the composite object to a field in BasicStatus.

Sounds good. I explored switching things around, but having `BasicStatus` contain the detailed information instead of the other way around makes some other things awkward. We can make if work, but let's see what they say.

File: cmd/dep/status.go (hunk 1 above)
Comment (zjs) [reply cmd/dep/status.go:174]: I was not. That seems like a better approach, but this pattern is copied from `func (out *tableOutput) BasicLine`. I think keepings things consistent is probably safest, and I think moving to using `text/tabwriter` probably belongs in a separate PR (or, at least, a separate commit).

File: cmd/dep/status.go (hunk 2 above)
Comment (zjs) [reply cmd/dep/status.go:1044]: I can see the appeal of that, but I think I'd like to leave this as-is for now; I think the right improvement here is to refactor this to remove some of the duplicated concepts (the 11 lines each of the `if`/`else` blocks seem way too similar to me, but I haven't figured out a good way to tidy that up).

**Review by akutz — APPROVED** [review]
